Implement Table.update()

eerovaher opened the pull request:
### Description
This pull request is meant to implement a `Table.update()` method, which would be analogous to Python's `dict.update()`, i.e. `t1.update(t2)` would add columns of `t2` to `t1` replacing columns if there is any overlap.

This pull request is opened as a draft to see if having such a method is considered to be worth the effort, and to get an idea how much effort that would be, but proper implementation also requires deciding about details such as whether the argument of the method should be a `Table` or if it could anything that can initialize a `Table` (e.g. `dict` or list of `Column` objects).

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will
review this pull request of some common things to look for. This list
is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] If the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

mhvk commented:
I quite like the idea. I guess a more minimal adjustment would be to allow `add_columns` to take a table as input. Though I find the analogy with `dict.update` quite compelling... 

ping @taldcroft

eerovaher commented:
Even if `Table.add_columns()` would accept a table as an argument it would still differ from `Table.update()` because it wouldn't allow columns to be replaced and it wouldn't merge table metadata.

mhvk commented:
It does allow columns to be replaced, but you're right about the metadata. The history of this is that Table originally used a structured `ndarray` for its storage, so the dict-like `update` was less obvious at the time. 

Of course, if we want to be a bit *more* dict-like, the following recent python we would add `__or__` and `__ior__`... (i.e., this would be `table1 |= table2`)...

eerovaher commented:
As far as I can tell column name conflicts cause `Table.add_columns()` to raise a `ValueError` if `rename_duplicate` is `False` or rename conflicting columns that are being added if it is `True`. But even if there is a way to make `add_columns()` replace columns that I am not aware of then the differences in metadata handling should still be enough to justify implementing `Table.update()`.

I see no reason why  `__or__` and `__ior__` shouldn't be implemented for `Table` objects similarly to how [PEP 584](https://www.python.org/dev/peps/pep-0584) implements them for dictionaries.

taldcroft commented on astropy/table/table.py:
```diff
@@ -3077,6 +3077,54 @@ def _replace_cols(self, columns):
 
         self.columns = columns
 
+    def update(self, other, column_conflicts='warn',
```
Since this is meant to be analogous to `dict.update`, I think we can make the default be `column_conflicts='silent'`.

eerovaher replied:
Should `metadata_conflicts='silent'` be the default too? That might be more consistent with `dict.update()` but less consistent with some of the table stacking operations.

mhvk replied:
I think so.

taldcroft commented on astropy/table/table.py:
```diff
@@ -3077,6 +3077,54 @@ def _replace_cols(self, columns):
+                else:
+                    warnings.warn('The input table already contains columns '
+                                  f'{common_cols}', AstropyUserWarning)
+        for col in other.colnames:
```
We can allow `other` to be anything dict-like, include `Table` using:
```
for name, col in other.items():
    self[name] = col
```

taldcroft commented on astropy/table/table.py:
```diff
@@ -3077,6 +3077,54 @@ def _replace_cols(self, columns):
+        """
+        Examples
+        --------
+        >>> t1 = Table({'a': ['foo', 'bar'], 'b': [True, False]})
```
I think just the one example is enough, where you add meta to each like below.

taldcroft reviewed (commented):
Good start, I like it. I thought a bit about `add_columns` but in the end agree with the reasoning from @eerovaher. Overloading this new functionality into that method would just make things confusing.

mhvk commented:
OK, I also am happy with just including `.update` - probably good to add `See Also` both for it and `add_columns`. Slightly larger question, to what level should one really minic `dict.update`? Its docstring is,
```
In [4]: dict.update?
Docstring:
D.update([E, ]**F) -> None.  Update D from dict/iterable E and F.
If E is present and has a .keys() method, then does:  for k in E: D[k] = E[k]
If E is present and lacks a .keys() method, then does:  for k, v in E: D[k] = v
In either case, this is followed by: for k in F:  D[k] = F[k]
```
Do we want the `**F` option? And the tuple one for `E`? (Note: am completely happy to keep it simple!)

eerovaher commented:
I would say that there are two simple options that are worth considering.
1) Only allow `other` to be a `Table`.
2) Allow `other` to be anything that can initialize a `Table`. In this case it would be possible to do something like
```python
if not isinstance(other, Table):
    other = Table(other)
```

mhvk commented:
@eerovaher - indeed, those are the real options (and are really what `dict.update` boils down to). To me, fine either way, to start with just table, or allow anything that initializes it.

taldcroft commented:
> start with just table, or allow anything that initializes it.

Sounds good. You probably want to initialize with `self.__class__(other)` instead of `Table(other)` though.

eerovaher commented:
The draft includes `column_conflicts` and `metadata_conflicts` as separate arguments. Would it be a good idea to replace them with a single `conflicts` argument? A third option would be to implement all three in such a way that `conflicts` would overrule the other two if it is specified.

mhvk commented:
@eerovaher - given that the operations have `metadata_conflicts`, I think keeping that name is best. I don't think a third argument is necessary. I'd go with either 2 or 0 extra arguments (0 since `.update` to me indicates clearly that one is overriding as needed).

taldcroft commented:
My first thought had been to eliminate both `conflicts` args, like @mhvk said for `dict.update`. I realized that this might be a limitation for use in the SkyCoord to_table workflow if you really want to be careful about existing columns. But I'm pretty much OK either way.

eerovaher commented:
Even though I have implemented the `column_conflicts` and `metadata_conflicts` arguments in the draft I am inclined to agree that removing them would make this method more similar to `dict.update()` and is therefore preferable.

eerovaher commented:
Getting code to work is indeed simple compared to getting it documented. 028dfe65ff8eb3b7ed2dbb7948f633a97ac7f5c7 already passed all the CI tests, but it took me two more commits to get the links in the See Also sections of the docstrings to work properly. Now this PR is really ready for final review.

mhvk reviewed (approved):
This is very nice, thanks also for taking care of the docs - they are indeed rather a pain, but also one of the things that make astropy good, so hopefully worth it! (I just wish the docs took less time to build and warnings were a bit easier to find...)

I'll approve but perhaps @taldcroft wants another look too.  If you can bring yourself to it, a rebase/squash to 1 or 2 commits would be nice.

taldcroft commented on astropy/table/table.py:
```diff
@@ -3077,6 +3084,91 @@ def _replace_cols(self, columns):
+             foo     1.0
+             bar     2.0
+
+        The updated columns are copied by default, but it is possible to link
```
I'd suggest cutting this example (sometimes less is better). It's not a common thing and the setup/demo of the functionality is a bit intricate. I think most people that need to use `copy=False` have an idea of how it works.

taldcroft commented on docs/changes/table/11904.feature.rst:
```diff
@@ -0,0 +1,2 @@
+It is now possible to perform dictionary-style updates on a ``Table`` using the
```
I'd suggest:
```
Added a new method ``Table.update()`` which does a dictionary-style
update of a ``Table`` by adding or replacing columns.
```

taldcroft reviewed (changes requested):
@eerovaher - thanks, this looks generally good (an in particular the actual code looks fine). I have a bunch of small suggestions on the testing and other bits.

eerovaher commented on astropy/table/tests/test_table.py:
```diff
@@ -1094,6 +1094,90 @@ def test_rename_columns(self, table_types):
             t.rename_columns(['a'], ['b', 'c'])
 
 
+@pytest.mark.usefixtures('table_types')
```
It's not testing `Table` and `QTable`, see the definition here: https://github.com/astropy/astropy/blob/ce3413c9e3aafb98c68b23e5164237d1a3968d4f/astropy/table/tests/conftest.py#L72
Should I still remove the decorator?

taldcroft replied:
Oops, my mistake, but the general comment still applies that the underlying table operations that get used in `update` are all well tested. So I think the answer is just to remove the decorator and use `Table` everywhere to create the tables.

eerovaher commented on astropy/table/operations.py:
```diff
@@ -684,6 +684,10 @@ def hstack(tables, join_type='outer',
 
+    See Also
+    --------
+    Table.add_columns, Table.replace_column, Table.update
```
Using `~` there breaks the build. Both `Table.add_columns` and `astropy.table.Table.add_columns` get converted to links that point to the correct location, but the links look exactly as written. I find this shorter notation less confusing.

taldcroft replied:
There's always a question of whether to prioritize the way the docs look as rendered in HTML or the plain text.

What's interesting is that numpy docs seem to have the best of both worlds. See e.g. https://numpy.org/doc/stable/reference/generated/numpy.empty.html#numpy.empty where the See Also links are clickable, but the inline docs look good as well:
```
See Also
--------
empty_like : Return an empty array with shape and type of input.
ones : Return a new array setting values to one.
zeros : Return a new array setting values to zero.
full : Return a new array of given shape filled with value.
```
On the whole our function doc strings do tend to use intersphinx links when possible, and especially for See Also it seems like a good thing to quickly be able to bring up the API docs for those items.

taldcroft replied:
And it looks like the astropy docs are doing some of this cleverness as well and the HTML docs already have correct intersphinx links. So never mind on my comment, and I learned something new about documentation writing!

eerovaher commented on astropy/table/tests/test_table.py:
```diff
@@ -1094,6 +1094,90 @@ def test_rename_columns(self, table_types):
+        assert np.all(t2['c'] == np.array([7, 8, 9]))
+        assert np.all(t2['d'] == np.array([7, 8, 7]))
+
+
```
The docstring contains an example with metadata merging that functions as a test, but a test with metadata conflicts is indeed missing.

eerovaher commented:
I have implemented the feedback, added a link from `Table.update()` documentation to metadata merging documentation, rebased the commits and squashed them down to one.

taldcroft commented on astropy/table/tests/test_table.py:
```diff
@@ -2192,6 +2192,100 @@ def test_kwargs_exception(self):
+class TestUpdate():
+
+    def _setup(self):
+        self.a = Column((1, 2, 3), name='a')
```
Awesome, great idea to just make these simple definitions which are exactly what you need.

taldcroft reviewed (approved):
@eerovaher - super, this looks great now. Thanks for making the changes which make the tests very much easier to understand and verify that they are doing the right thing.
